### PR TITLE
feat: add Maretron IPG100 TCP transport driver

### PR DESCRIPTION
**Pull request by sentientapp**
Adds a Maretron IPG100 TCP transport and `maretron-ipgjs` CLI for streaming 0xA5-framed N2K traffic as canboat plain CSV.

Includes CONNECT/SET_MODE handshake handling, outbound CSV/JSON PGN sends, reconnect behavior, and wire-format/parser tests.

 ## Details

  - This **does not consume** one of the IPG's licensed-client slots. Operators can 
    run multiple `MaretronIPGStream`/`maretron-ipgjs` connections alongside paid
    clients (N2KView).
  - The IPG100 caps total simultaneous client connections at 20 regardless of
    type, so these sessions still count toward that limit.
  - Binary mode is required: the IPG's default ASCII output is decoded using
    its internal PGN dictionary, which is limited.
  - The IPG100 does **no device-side PGN filtering** in either direction. Every
    frame on the bus reaches every connected client, and every frame any client
    writes is forwarded both to the bus and to all *other* connected clients.
  - The IPG handles fast-packet reassembly itself, so each frame carries a full
    logical N2K payload.
  - On TX, source address on the wire is always `0xFF` — the IPG substitutes
    its own claimed SA.

## Verification

  - Live-tested: signalk-server with this driver against an IPG100. Inbound PGNs
    flow into the SignalK data tree. Bidirectional confirmed both by signalk-server's
    startup ISO Requests (PGN 59904) reaching the bus and devices responding,
    and by `signalk-n2k-switching` round-tripping switch toggles.

## Summary by CodeRabbit

* **New Features**
  * Added support for Maretron IPG100 NMEA 2000 gateway.
  * Added a CLI tool to stream Maretron IPG100 N2K traffic.
  * Exposed Maretron IPG functionality from the library entrypoint.

* **Documentation**
  * Updated README with Maretron IPG100 and CLI usage example.

* **Tests**
  * Added comprehensive tests for Maretron IPG wire-format, stream behavior, reconnection, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/canboat/canboatjs/pull/430?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: lib/maretron-ipg.ts [hunk 1]
```diff
@@ -0,0 +1,802 @@
...
+ * token before matching, so the password is always wrapped in double
+ * quotes on the wire. A stock IPG with no configured password is matched
+ * by the literal two-character string `""`.
+ *
+ * The 4th token is a client-type label that the IPG parses but does not
+ * act on. Hard-coded to "MOBILE" to match the convention used elsewhere.
+ */
+export function buildConnectMessage(password: string): Buffer {
+  return Buffer.from(`CONNECT\t"${password}"\t\tMOBILE\0`, 'utf8')
+}
```
[the remark this answers is not in the document]
Comment (sentientapp) [reply lib/maretron-ipg.ts:283]: Fixed — added the regex guard in `buildConnectMessage` and a test that exercises each handshake-breaking character (`"`, `\t`, `\0`, `\r`, `\n`).

File: lib/maretron-ipg.ts [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+      .split(',')
+      .map((e: string) => e.trim())
+    jsonOutEvents.forEach((event: string) => {
+      options.app.on(event, (msg: PGN) => {
+        this.sendPGN(msg)
+        options.app.emit('connectionwrite', {
+          providerId: options.providerId
+        })
+      })
+    })
```
[the remark this answers is not in the document]
Comment (sentientapp) [reply lib/maretron-ipg.ts:406]: Pre-existing pattern across every transport in this repo (`ydgw02`, `actisense-serial`, `ikonvert` all do the same). Fixing it just here would be inconsistent; better as a separate project-wide pass. Leaving as-is for this PR.

File: lib/maretron-ipg.ts [hunk 3]
```diff
@@ -0,0 +1,802 @@
...
+    lastErr = err
+    this.setProviderError(err.message)
+  })
+
+  socket.on('close', () => {
+    this.debug('socket closed')
+    this.stopIdleTimer()
+    this.state = 'closed' as State
+    this.scheduleReconnect(lastErr)
+  })
```
[the remark this answers is not in the document]
Comment (sentientapp) [reply lib/maretron-ipg.ts:485]: Fixed — added an `authFailed` flag set in the `NO` handler and checked in the `'close'` handler before scheduling a reconnect. New test (`NO authentication failure does not schedule a reconnect or emit a fail-fast error`) confirms `reconnectTimer` stays null, no second `error` fires, and the factory isn't called again after advancing timers — also covers the nitpick about adding a reconnect regression for NO.

File: lib/maretron-ipg.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (sentientapp) [reply lib/maretron-ipg.ts:406]: Yes, please — that's a good way to capture it without blocking this PR. The scope is the unsubscribed `options.app.on(...)` handlers in `ydgw02` (3 handlers, including `ydFullRawOut`), `actisense-serial`, `ikonvert`, `w2k01`, `venus`, and this new `maretron-ipg` transport. Fix would be: capture each handler reference at registration time, remove them in the transport's `end()` cleanup path.

**Comment by sbender9** [comment]
Very cool! thanks!

**Comment by dirkwa** [comment]
Thanks a lot!
